sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala

```diff
@@ -1580,6 +1583,28 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
     assert(df.persist.take(1).apply(0).toSeq(100).asInstanceOf[Long] == 100)
   }
 
+  test("""SPARK-17154: df("column_name") should return correct result when we do self-join""") {
```

Comment:
What happens when:

    val joined = df.join(df, "inner")  // columns: col1, col2, col3, col1, col2, col3
    val selected  = joined.select(df("col1"))

As there are two plans with the same plan id, the breadth-first search will get one plan among them. So df("col") will be resolved. However, I think in this case, we should have an ambiguous error message.